### ISSUE TITLE
worker.recognize fails when run before previous job finishes

### ISSUE THREAD

**Issue by DR7777** [naptha/tesseract.js#875]
Hello!

I have just set up the OCR function and it seems to work if I use this (old deprecated):
`const result = await Tesseract.recognize(url, "eng");`

But does not work if I use this (new):
`const result = await worker.recognize(url);`

When I try to run it on a document the "new one" just randomly stops at certain pages / images and doesn't finish without throwing a bug.

See full code below. 
As I understood from the migration guide this is the only line I had to change?!

```
import Tesseract from "tesseract.js";

type OcrOnImagesOptions = {
  onProgress?: (progress: { current: number; total: number }) => void;
  onStart?: (progress: { current: 0; total: number }) => void;
};

const OcrOnImages = async (
  urls: string[],
  options?: OcrOnImagesOptions 
): Promise<
  Record<
    string,
    { text: string; boxes: Array<{ text: string; box: Tesseract.Box }> }
  >
> => {
  options.onStart && options.onStart({ current: 0, total: urls.length });
  const progress = { total: urls.length, current: 0 };

  const worker = await Tesseract.createWorker("eng");

  const promises = urls.map(async (url) => {
    try {
      // THIS DOES NOT WORK
      const result = await worker.recognize(url);
      console.log("result", result);

      // THIS WORKS PERFECTLY
      // const result = await Tesseract.recognize(url, "eng");
      // console.log("result1", result);

      progress.current += 1;
      options.onProgress && options.onProgress(progress);

      return {
        text: result.data.text,
        boxes: result.data.words.map((word) => ({
          text: word.text,
          box: word.bbox,
        })),
      };
    } catch (error) {
      console.error("Error processing URL:", url, error);
      return null;
    }
  });

  const results = await Promise.all(promises);

  return results.reduce((acc, data, index) => {
    return { ...acc, [index + 1]: data };
  }, {});
};

export default OcrOnImages;
```

Thanks so much for maintaining this library!

**Comment by Balearica** [comment]
Thanks for reporting.  I was able to replicate, and can confirm this appears to be a bug.  A minimal reproducible example is below.

```
const worker = await Tesseract.createWorker("eng");

for (let i=0; i<5; i++) {
  worker.recognize("https://raw.githubusercontent.com/naptha/tesseract.js/master/benchmarks/data/meditations.jpg").then(ret => {
    console.log(ret.data.text);
  })
}
```

Luckily, it appears to only happen under fairly niche conditions.  I was only able to replicate this behavior (1) when using a URL input and (2) calling `worker.recognize` again on the same worker without waiting for the previous job to finish.  In my example code above, changing `worker.recognize` to `await worker.recognize` waits until `worker.recognize` call completes before running the next one, which resolves the issue.  Therefore, the quick fix to your problem would be to always wait until `worker.recognize` is done running before running `worker.recognize` again. 

Regarding why this would not happen using `Tesseract.recognize`--that function creates a new worker *every time* it is called, so the same worker never gets two jobs.  However, this will cause other issues.  In addition to making a new worker for every job being inefficient, running `Tesseract.recognize` in parallel can lead to an uncontrollably large number of workers being spawned which can cause crashes due to resource constraints.  

If you want to run jobs in parallel, the best way to handle this is using schedulers.  Schedulers allow for using a defined number of workers in parallel to process jobs.  Schedulers are explained [here](https://github.com/naptha/tesseract.js/blob/master/docs/workers_vs_schedulers.md), and an example using schedulers is [here](https://github.com/naptha/tesseract.js/blob/master/examples/browser/basic-scheduler.html).

**Comment by DR7777** [comment]
Thanks for getting back so quick!

Actually I was awaiting the result. I just tried again like this and unfortunately I just keep waiting and waiting for the result of the second image...

```
const worker = await Tesseract.createWorker("eng");

  const promises = urls.map(async (url) => {
    try {
      const result = await worker.recognize(url);
      console.log("result", result);

      progress.current += 1;
      options.onProgress && options.onProgress(progress);

      return {
        text: result.data.text,
        boxes: result.data.words.map((word) => ({
          text: word.text,
          box: word.bbox,
        })),
      };
    } catch (error) {
      console.error("Error processing URL:", url, error);
      return null; 
    }
  });

  const results = await Promise.all(promises);
```


I also tried this method and it works fine and is quite fast aswell. But I assume starting a new worker and terminating it is quite inefficient? Even if I terminate it every time?

On my machine it works quite fast though:

```
  const promises = urls.map(async (url) => {
    try {
      const worker = await Tesseract.createWorker("eng");
      const result = await worker.recognize(url);
      console.log("result", result);

      progress.current += 1;
      options.onProgress && options.onProgress(progress);

      await worker.terminate();

      return {
        text: result.data.text,
        boxes: result.data.words.map((word) => ({
          text: word.text,
          box: word.bbox,
        })),
      };
    } catch (error) {
      console.error("Error processing URL:", url, error);
      return null; 
    }
  });

  const results = await Promise.all(promises);
```

**Comment by Balearica** [comment]
The root cause of this bug is that Tesseract.js workers only store one promise for each `action` at a time.  This can be seen in the code linked below.  If there is an unresolved promise with action `recognize`, and a new promise is created with action `recognize`, the new promise will overwrite the old promise.

https://github.com/naptha/tesseract.js/blob/master/src/createWorker.js#L51-L71

This causes the first `recognize` job that finishes to resolve the last promise created with action `recognize`.  The remaining jobs still run but do not return anything to the user, as the promise they are trying to resolve has already been resolved.

This should be easily fixable by making the identifier for all promises unique, which can be achieved by appending `jobId`.  I will do this in the next version of Tesseract.js.  However, even after this is fixed, I still do not recommend sending multiple `recognize` jobs to the same worker at the same time.  As described above, using a `scheduler` is the most performant way to handle executing multiple jobs in parallel.  Additionally, given that jobs have side effects and there is no queue system built into Tesseract.js workers, it may still be possible for sending a bunch of jobs to a single worker at the same time to cause issues.

**Comment by Balearica** [comment]
The fix described above has been implemented in the master branch and will be included in the next patch release (`v5.0.5`).
